presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,374 @@
+                    case IS_DISTINCT_FROM:
+                        return TRUE_LITERAL;
+                    default:
+                        throw new UnsupportedOperationException("not yet implemented");
+                }
+            }
+
+            Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
+            Type literalType = typeAnalyzer.getType(session, types, expression.getRight());
+
+            if (sourceType.equals(BIGINT) && literalType.equals(DOUBLE) ||
+                    sourceType.equals(BIGINT) && literalType.equals(REAL) ||
+                    sourceType.equals(INTEGER) && literalType.equals(REAL)) {
+                // Special case: implicit casts between BIGINT->DOUBLE/REAL and INTEGER->REAL are broken
+                // as they are not injective. Skip to avoid producing unexpected results
+                return expression;
+            }
+
+            if (!metadata.getTypeManager().canCoerce(sourceType, literalType)) {
+                // transformation is only valid when it's an implicit CAST
+                return expression;
+            }
+
+            Signature sourceToTarget = metadata.getFunctionRegistry().getCoercion(sourceType, literalType);
+
+            Optional<Type.Range> range = sourceType.getRange();
+            if (range.isPresent()) {
+                Object max = range.get().getMax();
+                Object maxInTargetType = functionInvoker.invoke(
+                        sourceToTarget,
+                        session.toConnectorSession(),
+                        max);
+
+                int upperBoundComparison = compare(literalType, literalValue, maxInTargetType);
+                if (upperBoundComparison > 0) {
+                    // larger than maximum representable value
+                    switch (operator) {
+                        case EQUAL:
+                        case GREATER_THAN:
+                        case GREATER_THAN_OR_EQUAL:
+                            return falseIfNotNull(cast.getExpression());
+                        case NOT_EQUAL:
+                        case LESS_THAN:
+                        case LESS_THAN_OR_EQUAL:
+                            return trueIfNotNull(cast.getExpression());
+                        case IS_DISTINCT_FROM:
+                            return TRUE_LITERAL;
+                        default:
+                            throw new UnsupportedOperationException("not yet implemented: " + operator);
+                    }
+                }
+
+                if (upperBoundComparison == 0) {
+                    // equal to max representable value
+                    switch (operator) {
+                        case GREATER_THAN:
+                            return falseIfNotNull(cast.getExpression());
+                        case GREATER_THAN_OR_EQUAL:
+                            return new ComparisonExpression(EQUAL, cast.getExpression(), literalEncoder.toExpression(max, sourceType));
```

Comment:
I'd add a comment to highlight that certain non-obvious assumption is being leveraged:
```
// implicit cast must maintain distinctness
```

Reply:
Also, coercion from decimal to double does not maintain distinctness (eg. values close to max value of `decimal(38,0)`).

Reply:
Good point. I missed that one. I'll add a check to bail out in that case.

Reply:
Can we do this while still maintaining the feature? i.e. can the check be based on _values_, not _types_?

Reply:
I'll make it bail out only for the cases where it can't work, but I will be allowed for decimals that fit entirely (just need to find that boundary)